### PR TITLE
DietPi-Software | Add DietPi Dashboard

### LINKED ISSUE #448: DietPi-Software | Web Dashboard for DietPi
### **ADMIN EDIT**

Dashboard is available for testing, see https://github.com/ravenclaw900/DietPi-Dashboard#installation for instructions.

#

Base:
- Must not require a webserver stack (implies self contained website). eg: Websockets and Bootstrap/JS
- Lightweight (minimal cpu and memory usage)

Main DietPi programs for web implementation:
- DietPi-Software
- DietPi-Config

Additional features:
- Terminal window

### PR DESCRIPTION
**Pull request by nonnorm**
It seems fitting that this gets software ID 200.

**Status**: Testing

**Reference**: https://github.com/MichaIng/DietPi/issues/448

**Commit list/description**:

- DietPi-Software | Add DietPi-Dashboard

### CODE REVIEW

File: dietpi/dietpi-software [hunk 1]
```diff
@@ -3499,6 +3507,41 @@ _EOF_
...
+			then
+				Download_Install "$(curl -sSfL 'https://api.github.com/repos/ravenclaw900/dietpi-dashboard/releases/latest' | mawk -F\" "/\"browser_download_url\": \".*dietpi-dashboard-$arch\"/{print \$4}")" /usr/local/bin/dietpi-dashboard
+			elif [[ $G_WHIP_RETURNED_VALUE == 'Nightly' ]]
+			then
+				Download_Install "https://nightly.link/ravenclaw900/DietPi-Dashboard/workflows/push-build/main/dietpi-dashboard-$arch.zip" dietpi-dashboard.zip
+				G_EXEC unzip -o dietpi-dashboard.zip
+				G_EXEC mv dietpi_dashboard /usr/local/bin/dietpi-dashboard
+			fi
+
+			G_EXEC chmod +x /usr/local/bin/dietpi-dashboard
```
Comment (MichaIng) [inline dietpi/dietpi-software]: When not shipping a service, a location in PATH is fine for easy access. But we want to ship a systemd unit, right? Then I suggest to install the binary to `/opt/dietpi-dashboard/dietpi-dashboard`, where we may add other files, like a readme, license, optional assets or such things, to have the option open.

**Comment by Joulinar** [comment]
btw, wich port do we use on the dashboard?

**Comment by ravenclaw900** [comment]
Currently 8080, but I plan to add a CLI and that could be changed.

**Comment by Joulinar** [comment]
yeah 8080 is already used by 3 other app. So we need to move definitely 😃

**Comment by ravenclaw900** [comment]
Hmm, when accessing terminal service ends with `dietpi-dashboard.service: Succeeded.`. No other logs, and it works fine when starting it manually

**Comment by MichaIng** [comment]
Hmm, does the original process stay active or does it spawn a child and ends somehow? This would explain the systemd service log, but not sure how it is related to accessing the terminal.

Basically the question is whether the PID stays the same all time or not.

**Comment by ravenclaw900** [comment]
It spawns a child, and the PID of the original service stays the same.

**Comment by MichaIng** [comment]
Only the original process matters, the child itself can start and end but the original/main process must remain active for default/`Type=simple` services. Does the following actually show an exit code or signal?
```sh
systemctl status dietpi-dashboard
```

**Comment by ravenclaw900** [comment]
No:
```
Nov 02 17:21:31 rpi4 dietpi-dashboard[3627147]: 2021-11-02 17:21:31,235 INFO  [dietpi_dashboard] Request to /ws/term
Nov 02 17:21:31 rpi4 systemd[1]: dietpi-dashboard.service: Succeeded.
Nov 02 17:21:31 rpi4 systemd[1]: dietpi-dashboard.service: Consumed 1.167s CPU time.
```

**Comment by MichaIng** [comment]
Mysterious. I'll try to replicate. Probably starting the binary via `strace` helps to get all details about what happens.

File: dietpi/dietpi-software [hunk 2]
```diff
@@ -3551,6 +3551,15 @@ _EOF_
 
 			G_EXEC chmod +x /opt/dietpi-dashboard/dietpi-dashboard
 
+			cat << _EOF_ > /opt/dietpi-dashboard/config.toml
+port = 8080
```
Comment (MichaIng) [inline dietpi/dietpi-software]: This port is in use already by Airsonic. To avoid conflicts, we need to use a different one, the next free one is `8088`. Also, we should not overwrite the config file if it exists already, to preserve user adjustments:
```sh
[[ -f '/opt/dietpi-dashboard/config.toml' ]] || cat << _EOF_ > /opt/dietpi-dashboard/config.toml
```

Many thanks for adding this, awesome! 👍

Comment (nonnorm) [reply dietpi/dietpi-software]: 8088 seems good, I'll set it as the default as well when nothing is specifically set.

File: dietpi/dietpi-software [hunk 3]
```diff
@@ -3508,6 +3517,74 @@ _EOF_
...
+				Download_Install "$(curl -sSfL 'https://api.github.com/repos/ravenclaw900/dietpi-dashboard/releases/latest' | mawk -F\" "/\"browser_download_url\": \".*dietpi-dashboard-$arch\"/{print \$4}")" /opt/dietpi-dashboard/dietpi-dashboard
+			elif [[ $G_WHIP_RETURNED_VALUE == 'Nightly' ]]
+			then
+				no_check_url=1 Download_Install "https://nightly.link/ravenclaw900/DietPi-Dashboard/workflows/push-build/main/dietpi-dashboard-$arch.zip" dietpi-dashboard
+				G_EXEC mv dietpi-dashboard/dietpi_dashboard /opt/dietpi-dashboard/dietpi-dashboard
+			fi
+
+			G_EXEC chmod +x /opt/dietpi-dashboard/dietpi-dashboard
+
+			cat << _EOF_ > /opt/dietpi-dashboard/config.toml
```
Comment (MichaIng) [inline dietpi/dietpi-software]: Shall we keep the convention to have any configs (and data) in `/mnt/dietpi_userdata/<name>`? But for now it is probably overkill with then a single file only in each dir, and it would require an additional option to define the config file path... We can move this any time later when more features which write data, or split config files are implemented.

But we shouldn't overwrite existing config files:
```sh
[[ -f '/opt/dietpi-dashboard/config.toml' ]] || cat << _EOF_ > /opt/dietpi-dashboard/config.toml
```

Not a blocker at all, just an additional little hint for `<<` here documents (same below for the systemd unit): If the ending string `_EOF_` is wrapped into single quotes, the here document is taken literally, hence no variables or command substitutions are expanded. I started to hence use them whenever there is no variable/command substitution intended, similarly how I wrap all literal strings in single quotes instead of double quotes, just to be more explicit:
```sh
cat << '_EOF_' > /opt/dietpi-dashboard/config.toml
```

File: dietpi/dietpi-software [hunk 4]
```diff
@@ -13658,6 +13735,20 @@ _EOF_
...
+		if (( ${aSOFTWARE_INSTALL_STATE[$software_id]} == -1 )); then
+
+			Banner_Uninstalling
+			[[ -d '/opt/dietpi-dashboard' ]] && G_EXEC rm -R /opt/dietpi-dashboard
+			if [[ -f '/etc/systemd/system/dietpi-dashboard.service' ]]; then
+
+				G_EXEC systemctl disable --now dietpi-dashboard
+				G_EXEC rm /etc/systemd/system/dietpi-dashboard.service
+
+			fi
```
Comment (MichaIng) [inline dietpi/dietpi-software:13733]: In case `dietpi-services` was used to change process priorities, or a manual override was created:
```sh
[[ -d '/etc/systemd/system/dietpi-dashboard.service.d' ]] && G_EXEC rm -R /etc/systemd/system/dietpi-dashboard.service.d
```
And I personally prefer the service to be disabled/removed first and the executable afterwards. It doesn't make a difference here, but is the better convention to cover cases where the service contains an `ExecStop`, which would then fail.

File: dietpi/dietpi-software [hunk 5]
```diff
@@ -1199,6 +1199,15 @@ INDEX_BROWSER_TARGET=$INDEX_BROWSER_TARGET"
 		aSOFTWARE_CATX[$software_id]=8
 		aSOFTWARE_DOCS[$software_id]='https://dietpi.com/docs/software/cloud/#k3s'
 
+		#------------------
+		software_id=200
```
Comment (MichaIng) [inline dietpi/dietpi-software:1203]: Btw, `142` would be free (due to CouchPotato removal with v7.7). But actually I like this using the round number 🙂.

**Comment by Joulinar** [comment]
Maybe we should use a special port number as well like 8888 🤣

**Comment by MichaIng** [comment]
`8888` is used by RPi-Monitor already: https://dietpi.com/docs/software/system_stats/#rpi-monitor

Too beautiful numbers have a higher change of conflicts, so IMO `8088` is a good compromise, still beautiful enough 🙂.

File: dietpi/dietpi-software [hunk 6]
```diff
@@ -1199,6 +1199,15 @@ INDEX_BROWSER_TARGET=$INDEX_BROWSER_TARGET"
...
+
+		aSOFTWARE_NAME[$software_id]='DietPi Dashboard (beta)'
+		aSOFTWARE_DESC[$software_id]='Official DietPi web dashboard'
+		aSOFTWARE_CATX[$software_id]=8
+		aSOFTWARE_DOCS[$software_id]='https://dietpi.com/docs/dietpi_tools/#dietpi-dashboard'
+		aSOFTWARE_INTERACTIVE[$software_id]=1
```
Comment (MichaIng) [inline dietpi/dietpi-software]: Shall we allow non-interactive installs which then default to `Stable`? Also we could add a `dietpi.txt` setting for this.

File: dietpi/dietpi-software [hunk 7]
```diff
@@ -3508,6 +3517,74 @@ _EOF_
...
+				no_check_url=1 Download_Install "https://nightly.link/ravenclaw900/DietPi-Dashboard/workflows/push-build/main/dietpi-dashboard-$arch.zip" dietpi-dashboard
+				G_EXEC mv dietpi-dashboard/dietpi_dashboard /opt/dietpi-dashboard/dietpi-dashboard
+			fi
+
+			G_EXEC chmod +x /opt/dietpi-dashboard/dietpi-dashboard
+
+			cat << _EOF_ > /opt/dietpi-dashboard/config.toml
+port = 8088
+
+[tls]
```
Comment (MichaIng) [inline dietpi/dietpi-software]: Sections are nice to structure a config file, on the other hand it makes it more difficult to script config changes. E.g. if there are several section with `enable`, then `grep`/`sed`/`G_CONFIG_INJECT` cannot be easily used anymore to get/set settings.

Probably there are other arguments (e.g. Rust crate wise) to keep using sections, but otherwise I suggest to skip it for now and use instead:
```
tls = false
cert = "/path/to/cert"
key = "/path/to/key"
```
Hence keep using unique keys and structure the config file, if wanted, with comments.

File: dietpi/dietpi-software (hunk 6 above)
Comment (nonnorm) [reply dietpi/dietpi-software]: Sounds good. Would we do this with the `$G_INTERACTIVE` variable? The option will also probably be removed once the dashboard is out of beta.

Comment (MichaIng) [reply dietpi/dietpi-software]: Our G_WHIPs are cancelled automatically in this case. We could initiate the variable with the setting in `dietpi.txt`, if it is empty assign `Stable`, then show the whiptail and if it is returns a selection override the variable again. So priority is: Whiptail selection > `dietpi.txt` > `Stable`

File: dietpi/dietpi-software [hunk 8]
```diff
@@ -3508,6 +3516,79 @@ _EOF_
...
+			local version=$(sed -n '/^[[:blank:]]*SOFTWARE_DIETPI_DASHBOARD_VERSION=/{s/^[^=]*=//p;q}' /boot/dietpi.txt)
+			[[ $version ]] || version=Stable
+
+			G_WHIP_MENU_ARRAY=(
+					'Stable' ': Use a release version of the DietPi Dashboard.'
+					'Nightly' ': Use an unstable version of the DietPi Dashboard.'
+			)
+
+			local choice_required=
+			while :
```
Comment (MichaIng) [inline dietpi/dietpi-software]: The loop needs to be removed then, the following is sufficient to leave the variable untouched when cancelled (or non-interactive):
```sh
G_WHIP_MENU "${choice_required}Please choose which version of DietPi Dashboard should be installed." && version=$G_WHIP_RETURNED_VALUE
```
Probably we could add:
```
G_WHIP_BUTTON_CANCEL_TEXT=$version
G_WHIP_DEFAULT_ITEM=$version
```
to pre-select the dietpi.txt value/default and replace the "Cancel" button text with the resulting selection when hitting it 🤔.

**Comment by ravenclaw900** [comment]
Do you suggest removing the dashboard from the DietPi controlled services, because it can be used to install software, and we probably don't want it stopping then.

**Comment by MichaIng** [comment]
Good point! Memory usage is sufficiently low to ignore this during installs and such. It then needs to be enabled:
```sh
systemctl enable --now dietpi-dashboard
```
or
```sh
systemctl enable dietpi-dashboard
aSTART_SERVICES+=('dietpi-dashboard')
```
so that it is started when `dietpi-software` finishes.

File: dietpi/dietpi-services [hunk 9]
```diff
@@ -217,23 +217,24 @@ Available services:
 
 			# SSH
 			aSERVICE_NAME+=('dropbear')
-			aSERVICE_NAME+=('ssh') 				# OpenSSH Server
+			aSERVICE_NAME+=('ssh') 						# OpenSSH Server
```
Comment (MichaIng) [inline dietpi/dietpi-services]: When a tab is shown like 8 spaces, all is messed up now 😄. Please revert the changes here as we have spacing aligned for 1 tab = 8 spaces on all scripts.

File: dietpi/dietpi-software [hunk 10]
```diff
@@ -1199,6 +1199,14 @@ INDEX_BROWSER_TARGET=$INDEX_BROWSER_TARGET"
 		aSOFTWARE_CATX[$software_id]=8
 		aSOFTWARE_DOCS[$software_id]='https://dietpi.com/docs/software/cloud/#k3s'
 
+		#------------------
+		software_id=200
+
+		aSOFTWARE_NAME[$software_id]='DietPi Dashboard (beta)'
```
Comment (MichaIng) [inline dietpi/dietpi-software]: Another question: Shall we call it "DietPi-Dashboard" with a dash before "Dash", to align it with other script names, like "DietPi-CloudShell" and "DietPi-VPN"?

Comment (nonnorm) [reply dietpi/dietpi-software]: Yeah, that will probably match things up better.

**Comment by MichaIng** [comment]
#### Testing
- [ ] x86_64 Stretch
- [ ] x86_64 Buster
- [ ] x86_64 Bullseye
- [ ] ARMv6 Stretch
- [ ] ARMv6 Buster
- [ ] ARMv6 Bullseye
- [ ] ARMv7 Stretch
- [ ] ARMv7 Buster
- [ ] ARMv7 Bullseye
- [ ] ARMv8 Stretch
- [ ] ARMv8 Buster
- [ ] ARMv8 Bullseye

**Comment by MichaIng** [comment]
Okay generally it works fine, with both binary types musl and gnu, on all architectures and distro versions. Remaining issues will be tracked at the dashboard repo, but do not (necessarily) a change in the install code.

**Comment by Dulani** [comment]
Port 8088 _was_ certainly a 'beautiful' choice (I figured it out eventually), but your [documentation](https://dietpi.com/docs/software/system_stats/#:~:text=Access%20the%20web%20interface) still says the default port is: `5252`.  Suggest updating that.

**Comment by Joulinar** [comment]
Port 5252 is the new port as we hat a port conflict on 8088. Therefore docs are correct. You need to update to latest DietPi version and install latest nightly build for the Dashboard to get 5252 activate.
